File: src/types/analysis.ts
export type RGB = {
  r: number;
  g: number;
  b: number;
};

export type ExtractedColor = {
  hex: string;
  rgb: RGB;
  weight: number;
  role: "primary" | "secondary" | "accent" | "neutral";
};

export type ColorMatch = {
  reference: ExtractedColor;
  target: ExtractedColor | null;
  distance: number;
};

export type AnalysisResult = {
  referencePalette: ExtractedColor[];
  targetPalette: ExtractedColor[];
  colorMatchScore: number;
  findings: string[];
  suggestions: string[];
  matches: ColorMatch[];
  confidence: number;
  summary: string;
  createdAt: string;
};

File: backend/app/models/analysis.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict


class RGB(BaseModel):
    r: int
    g: int
    b: int


class ExtractedColor(BaseModel):
    hex: str
    rgb: RGB
    weight: float
    role: Literal["primary", "secondary", "accent", "neutral"]


class ColorMatch(BaseModel):
    reference: ExtractedColor
    target: ExtractedColor | None
    distance: float


class AnalysisResult(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    referencePalette: list[ExtractedColor]
    targetPalette: list[ExtractedColor]
    colorMatchScore: int
    findings: list[str]
    suggestions: list[str]
    matches: list[ColorMatch]
    confidence: float
    summary: str
    createdAt: str
